qualtran/bloqs/factoring/mod_add.py

```diff
@@ -138,3 +141,103 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
 
     def t_complexity(self) -> 'TComplexity':
         return TComplexity(t=2 * self.bitsize)
+
+
+@frozen
+class MontgomeryModAdd(Bloq):
+    r"""An n-bit modular addition gate.
+
+    This gate is designed to operate on integers in the Montgomery form.
+    Implements |x>|y> => |x>|y + x % p> using $4n$ Toffoli
+    gates.
+
+    Args:
+        bitsize: Number of bits used to represent each integer.
+        p: The modulus for the addition.
+
+    Registers:
+        x: A bitsize-sized input register (register x above).
+        y: A bitsize-sized input/output register (register y above).
+
+    References:
+        [How to compute a 256-bit elliptic curve private key with only 50 million Toffoli gates](https://arxiv.org/abs/2306.08585) Fig 6a and 8
+    """
+
+    bitsize: int
+    p: int
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature([Register('x', bitsize=self.bitsize), Register('y', bitsize=self.bitsize)])
+
+    def on_classical_vals(
+        self, x: 'ClassicalValT', y: 'ClassicalValT'
+    ) -> Dict[str, 'ClassicalValT']:
+
+        y += x
+        y -= self.p
+
+        if y < 0:
+            y += self.p
+
+        return {'x': x, 'y': y}
+
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', x: SoquetT, y: SoquetT
+    ) -> Dict[str, 'SoquetT']:
+
+        # Allocate ancilla bits for use in addition.
+        junk_bit = bb.allocate(n=1)
+        sign = bb.allocate(n=1)
+
+        # Join ancilla bits to x and y registers in order to be able to compute addition of
+        # bitsize+1 registers. This allows us to keep track of the sign of the y register after a
+        # constant subtraction circuit.
+        x_split = bb.split(x)
+        y_split = bb.split(y)
+        x = bb.join(np.concatenate([[junk_bit], x_split]))
+        y = bb.join(np.concatenate([[sign], y_split]))
+
+        # Perform in-place addition on quantum register y.
+        x, y = bb.add(Add(bitsize=self.bitsize + 1), a=x, b=y)
+
+        # Temporary solution to equalize the bitlength of the x and y registers for Add().
```

Comment:
Can we change the signature of `Add` to support `a_bitsize` and `b_bitsize`; update tensor decomposition and open an issue that tracks updating the decomposition? So that the costs and tensor simulations / classical simulations work for aribtrary bitsizes but the decomposition works only when bitsizes are the same and that can be updated later? 

I think it's much harder to find such TODOs than fixing the decomposition later.

Reply:
I'm going to say this is outside of the scope of the current PR